### PR TITLE
fixes #12642

### LINKED ISSUE #12642: If a var holding a proc is assigned from a case statement, its "nosideeffect" state is set to that of the first arm of case statement
If a proc-holding var is assigned a value using a `case` statement, it's "noSideEffect" state is set to the one from the first arm of the case statement. i.e. if the second or later arm of the `case` has a proc with side effects, it will fail compilation.

So can one of these be made possible:
- Force the "side effect state" of a proc even if it actually has no side effects 
- Force the "side effect state" of a proc holding var
- If a var is assigned from a case statement as in this example, its "side effect state" should be "**has** side effects" irrespective of the order of the case statement cases.

### Example `t12642.nim`
https://play.nim-lang.org/#ix=21tP

```nim
# fails ->  nim c t12642.nim
# passes -> nim c -d:workaround t12642.nim

let
  choice = 0

proc withoutSideEffects(): int = 0
proc withSideEffects(): int = echo "foo" # the echo causes the side effect

when defined(workaround):
  # Workaround is to put the proc *without* side effects at the end of the
  # case statement.
  let procPtr = case choice
                of 0: withSideEffects
                else: withoutSideEffects
else:
  let procPtr = case choice
                of 0: withoutSideEffects
                else: withSideEffects

echo procPtr.repr
```

Run `nim c t12642.nim`.

### Current Output

> Error: type mismatch: got <proc (): int{.gcsafe, locks: 0.}> but expected 'proc (): int{.noSideEffect, gcsafe, locks: 0.}'

### Expected Output

_Should have printed the addr of the `procPtr`._

### Workaround

Re-arrange the case statement so that the first case is the one where the proc **with side effects** is assigned,

But this workaround is not very robust, because tomorrow if the `withSideEffects` proc gets refactored to have no side effects, the case statement might need to be shuffled again.

So running `nim c -d:workaround t12642.nim` will work.

### Additional Information

- https://stackoverflow.com/questions/53726458/nim-aliasing-procedures-with-side-effects

```
$ nim -v
Nim Compiler Version 1.1.1 [Linux: amd64]
Compiled at 2019-11-11
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: 7e689873e203620c3a14f1697129cd0e3fd548e4
active boot switches: -d:release
```

---

**Update (2021/06/10)**:

I am re-confirming this issue on the latest devel:

```
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-06-10
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 19918ceb2b099440cd3e43b6d7fff313de6a84a2
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by Araq**

### CODE REVIEW

File: testament/categories.nim [hunk 1]
```diff
@@ -459,11 +459,12 @@ proc testNimblePackages(r: var TResults; cat: Category; packageFilter: string) =
       r.addResult(packageFileTest, targetC, "", "", reBuildFailed)
 
   except JsonParsingError:
-    echo "[Warning] - Cannot run nimble tests: Invalid package file."
+    errors = 1
```
Comment (ringabout) [inline testament/categories.nim:462]:
```suggestion
    echo "[Warning] - Cannot run nimble tests: Invalid package file."
```

Comment (Araq) [reply testament/categories.nim:462]: But that's taken from your PR.

Comment (ringabout) [reply testament/categories.nim:462]: My PR raises exception at the former situation, namely https://github.com/nim-lang/Nim/pull/18811/files#r702753638

image

File: testament/categories.nim [hunk 2]
```diff
@@ -459,11 +459,12 @@ proc testNimblePackages(r: var TResults; cat: Category; packageFilter: string) =
...
     r.addResult(packageFileTest, targetC, "", "Invalid package file", reBuildFailed)
```
Comment (ringabout) [inline testament/categories.nim:463]:
```suggestion
    r.addResult(packageFileTest, targetC, "", "Invalid package file", reBuildFailed)
    raise
```

**Pull request merged**